run-an-ethereum-node-with-nimbus

cheny0 opened the pull request:
This draft is first for Campo to review.

cheny0 committed:
Final draft of the quickstart guide

cheny0 committed:
Final draft of the quickstart guide

cheny0 committed:
Fix merge conflicts

tersec committed:
fix EIP4844 withBlck (#4411)

* fix EIP4844 withBlck

* don't raiseAssert by default

tersec committed:
add database beaconstate tests for capella and eip4844 (#4429)

tersec committed:
use v1.3.0-alpha.2 EF consensus spec test vectors (#4430)

tersec committed:
add EIP4844 fork choice tests (#4431)

tersec committed:
update EF consensus spec URLs from v1.3.0-alpha.1 to v1.3.0-alpha.2 (#4432)

tersec committed:
consolidate consensus spec transition test fixtures (#4428)

* consolidate consensus spec transition test fixtures

* include capella

* consoliate fork test fixtures

* note change in EIP-4844 process_block in alpha.2

arnetheduck committed:
fix capella+ summary loading (#4433)

...once and for all.

arnetheduck committed:
60% state replay speedup (#4434)

* 60% state replay speedup

* don't use HashList for epoch participation - in addition to the code
currently clearing the caches several times redundantly, clearing has to
be done each block nullifying the benefit (35%)
* introduce active balance cache - computing it is slow due to cache
unfriendliness in the random access pattern and bounds checking and we
do it for every block - this cache follows the same update pattern as
the active validator index cache (20%)
* avoid recomputing base reward several times per attestation (5%)

Applying 1024 blocks goes from 20s to ~8s on my laptop - these kinds of
requests happen on historical REST queries but also whenever there's a
reorg.

* fix test and diffs

arnetheduck committed:
Implement skip_randao_verification for blinded blocks (#4435)

* Implement skip_randao_verification for blinded blocks

* fix redundant randao verification on block replay (5% faster)
* check randao in REST instead of internally
* avoid redundant copies when making blocks
* cleanup leftover randao skipping code

* fix test summary

zah committed:
Publish official docker images for the Nimbus validator client (#4439)

Fixes https://github.com/status-im/nimbus-eth2/issues/4426

zah committed:
Migrating the deposit contract snapshot can no longer fail on start-up (#4438)

The missing piece of data that had to be obtained previously from
the configured EL client is now part of the network metadata baked
into the binary.

etan-status committed:
validate EL block hash in EL simulation (#4420)

When simulating EL with `--optimistic` flag, perform block hash check.

zah commented on docs/the_nimbus_book/updates/run-an-ethereum-node-with-nimbus.md:
```diff
@@ -0,0 +1,469 @@
+
+## Prerequisites
+
+Before you start, first learn the basics of [Ethereum :octicons-tab-external-16:](https://ethereum.org/en/developers/docs/intro-to-ethereum/){:target="_blank"} and how to command line. The guide assumes you run both clients on the same machine using Ubuntu Linix. Here are more detailed requirements for your system:
```
The phrase "how to command line" seems strange to me. Perhaps it was a typo?

cheny0 replied:
Should be "how to use the command line"

zah commented on docs/the_nimbus_book/updates/run-an-ethereum-node-with-nimbus.md:
````diff
@@ -0,0 +1,469 @@
+
+1. Use the `mkdir` command to create a directory with the name `nimbus-eth2` to hold Nimbus data and applications.  
+```sh
+mkdir -p nimbus-eth2
````
Let's recommend using `/var/lib/nimbus` which is the directory used in the Nimbus package and the SomerEsat guide. This will make sure people can upgrade smoothly to the Nimbus package in the future.

cheny0 replied:
Ok

zah commented on docs/the_nimbus_book/updates/run-an-ethereum-node-with-nimbus.md:
````diff
@@ -0,0 +1,469 @@
+mkdir -p nimbus-eth2
+```
+
+1. Download the latest [Nimbus release :octicons-tab-external-16:](https://github.com/status-im/nimbus-eth2/releases){:target="_blank"} using the `wget` command. In the following command, replace `<version_number>` with the version number of the latest release, replace `<name_of_the_release_package>` with the name of the package to download. You need to download the `tar.gz` file with "Linix_arm6 ...
````
Does `mkdocs` support any form of pre-processing of the md files? (i.e. replacing placeholders in the text with values specified in a configuration file or the command line).

It would be nice to insert the latest version of the software automatically here.

zah commented on docs/the_nimbus_book/updates/run-an-ethereum-node-with-nimbus.md:
````diff
@@ -0,0 +1,469 @@
+
+    === "Mainnet"
+        ```sh
+        ./run-mainnet-beacon-node.sh
````
The suggestion to use these scripts has become a bit problematic after the merge, because they don't specify a `--web3-url` by default. The client cannot work properly when launched without a `--web3-url`.

cheny0 replied:
> The suggestion to use these scripts has become a bit problematic after the merge, because they don't specify a `--web3-url` by default. The client cannot work properly when launched without a `--web3-url`.

Ok. I just noticed in the current guide this has been changed to `./run-mainnet-beacon-node.sh --web3-url=http://127.0.0.1:8551 --jwt-secret=/tmp/jwtsecret`.

zah commented on docs/the_nimbus_book/updates/run-an-ethereum-node-with-nimbus.md:
````diff
@@ -0,0 +1,469 @@
+    Nimbus automatically begins to look for peers and synchronize data from them, you can see log messages like this:
+
+    ```
+    INF 2022-06-16 13:23:11.008+02:00 Slot start
````
This looks like the Chronicles' `textblocks` output format, which is not currently used in the default configuration of Nimbus.

cheny0 replied:
> This looks like the Chronicles' `textblocks` output format, which is not currently used in the default configuration of Nimbus.

<img width="512" alt="image" src="https://user-images.githubusercontent.com/67916832/209919701-e333023e-4b03-4659-baa4-3b183aedf420.png">
You mean the texts are in one single line now? This is what I see when I start Nimbus.

arnetheduck replied:
I think this was manually reformatted at some point for clarity since there's no word wrap

zah commented on docs/the_nimbus_book/updates/run-an-ethereum-node-with-nimbus.md:
```diff
@@ -0,0 +1,469 @@
+    | `peers` | The number of peers Nimbus is currently connected to. |
+    | `head` | The most recent block synced to so far. In this example, `5d59aba3` is the first part of the block hash, `4021234` is the slot number. |
+    | `finalized` | The most recent finalized epoch synced to so far. In this example, `125661` is the epoch, `82616f78` is the checkpoint hash. |
+    | `delay` | WAITING FOR HELPDESK REPLY |
```
This may indicate that the client was busy doing some computational work which forced it to execute the slot processing code with a certain delay (indicated here).

cheny0 replied:
Got it

cheny0 replied:
> This may indicate that the client was busy doing some computational work which forced it to execute the slot processing code with a certain delay (indicated here).

Got it

zah commented on docs/the_nimbus_book/updates/run-an-ethereum-node-with-nimbus.md:
````diff
@@ -0,0 +1,469 @@
+        U - updating peer's status information
+    ```
+
+    Data synchronization may take hours to a few days.
````
This sentence seems a bit unnatural to me (purely when it comes to use of English).

zah commented on docs/the_nimbus_book/updates/run-an-ethereum-node-with-nimbus.md:
```diff
@@ -0,0 +1,469 @@
+
+## Run an execution client
+
+Along with the consensus client, you need to set up an execution client. Execution clients are for processing and broadcasting transactions and managing Ethereum's state. We tested Nimbus with all [major execution clients :octicons-tab-external-16:](https://ethereum.org/en/developers/docs/nodes-and-clients/#execution-clients){:target="_blank"}, you can choose one that suits your needs.
```
How about "Nimbus is tested with all ..."

cheny0 replied:
Ok

cheny0 replied:
> How about "Nimbus is tested with all ..."

Ok

zah commented on docs/the_nimbus_book/updates/run-an-ethereum-node-with-nimbus.md:
```diff
@@ -0,0 +1,469 @@
+
+Along with the consensus client, you need to set up an execution client. Execution clients are for processing and broadcasting transactions and managing Ethereum's state. We tested Nimbus with all [major execution clients :octicons-tab-external-16:](https://ethereum.org/en/developers/docs/nodes-and-clients/#execution-clients){:target="_blank"}, you can choose one that suits your needs.
+
+In parallel to the [Nimbus consensus client :octicons-tab-external-16:](https://github.com/status-im/nimbus-eth2){:target="_blank"}, we are working hard on an execution client. While this project is still in development, we welcome you to experiment with it. To learn how the execution client works, check out its [GitHub repository :octicons-tab-external-16:](https://github.com/status-im/nimbus-et ...
```
It's probably a bit premature to make this recommendation.

cheny0 replied:
I can remove it, but this part already exists in the current guide

cheny0 replied:
> It's probably a bit premature to make this recommendation.

I can remove it, but this part already exists in the current guide

zah commented on docs/the_nimbus_book/updates/run-an-ethereum-node-with-nimbus.md:
```diff
@@ -0,0 +1,469 @@
+
+In parallel to the [Nimbus consensus client :octicons-tab-external-16:](https://github.com/status-im/nimbus-eth2){:target="_blank"}, we are working hard on an execution client. While this project is still in development, we welcome you to experiment with it. To learn how the execution client works, check out its [GitHub repository :octicons-tab-external-16:](https://github.com/status-im/nimbus-et ...
+
+=== "Besu"
```
I'm concerned that these instructions for running third-party clients may get a out of date in the future, without us noticing.

arnetheduck replied:
I think it makes sense to include "basic" instructions, ie the minimal set of flags needed to run things - these are unlikely to change - then, for every client, also include a link to the source documentation of _their_ "getting started" guide.

Keeping things up to date is a matter of scheduling a review of the guide every 6 months and run through it - it makes sense regardless because even "our" recommendations change.

cheny0 replied:
I'll try to minimize the content here.

etan-status committed:
bump `nim-eth`, extend empty block fallback for EIP4844 (#4425)

Implements `emptyPayloadToBlockHeader` for EIP-4844.
https://github.com/status-im/nim-eth/pull/570

zah committed:
Documentation for the --with-deposit-snapshot option (#4440)

zah committed:
Don't package the nimbus_validator_client in the nimbus-eth2 docker images; Document the new dedicated image (#4441)

zah committed:
Version 22.12.0

etan-status committed:
fix `produceBlockV2` call for Capella (#4443)

Avoid issue when calling `/eth/v2/validator/blocks/{slot}` for a
Capella slot via VC, same logic as in `validator_duties`.

etan-status committed:
display owl on Capella transition (#4442)

Trigger ANSI art on upgrade to Capella, similar to the merge.
Future extension could log blinking art when user successfully managed
to get BLS to Execution change included into a block for a validator.

Art created by http://beatscribe.com/ (beatscribe#1008 on Discord)

arnetheduck committed:
Backfill only up to MIN_EPOCHS_FOR_BLOCK_REQUESTS blocks (#4421)

When backfilling, we only need to download blocks that are newer than
MIN_EPOCHS_FOR_BLOCK_REQUESTS - the rest cannot reliably be fetched from
the network and does not have to be provided to others.

This change affects only trusted-node-synced clients - genesis sync
continues to work as before (because it needs to construct a state by
building it from genesis).

Those wishing to complete a backfill should do so with era files
instead.

arnetheduck committed:
fix countdown over time period (#4449)

*sigh*

fixes:
```
Error: unhandled exception:
beacon_chain/beacon_chain_db_light_client.nim(214, 12)
`period.isSupportedBySQLite`  [AssertionDefect]
```

tersec committed:
fix getForkedBlock to read EIP4844 and Capella blocks (#4450)

* fix getForkedBlock to read EIP4844 and Capella blocks

* update geth develop URLs

omahs committed:
Fix: minor typo (#4452)

Fix: minor typo

tersec committed:
manual rebase of Bump wheel from 0.37.1 to 0.38.1 in /docs #4446 (#4456)

tersec committed:
capella validator withdrawal credentials aren't immutable (#4455)

arnetheduck committed:
Allow reindexing to start from an arbitrary state (#4437)

When not backfilling all the way to genesis (#4421), it becomes more
useful to start rebuilding the historical indices from an arbitrary
starting point.

To rebuild the index from non-genesis, a state and an unbroken block
history is needed - here, we allow loading the state from an era file
and recreating the history from there onwards.

* speed up partial era state loading

henridf committed:
eip4844 gossip (#4444)

* eip4844 gossip

* Check BLSFieldElement range validity in gossip validation

* lint/nits cleanup

* Use template to avoid an assignment with copy.

* More review feedback

* lint

* lint

* processSignedBeaconBlockAndBlobsSidecar: clean up error handling flow

* Undo factoring-out of beacon blocks validator installation

arnetheduck committed:
Always-on optimistic mode (#4458)

With https://github.com/status-im/nimbus-eth2/pull/4420 implemented, the
checks that we perform are equivalent to those of a `SYNCING` EL - as
such, we can treat missing EL the same as SYNCING and proceed with an
optimistic sync.

This mode of operation significantly speeds up recovery after an offline
EL event because the CL is already synced and can immediately inform the
EL of the latest head.

It also allows using a beacon node for consensus archival queries
without an execution client.

* deprecate `--optimistic` flag
* log block details on EL error, soften log level because we can now
continue to operate
* `UnviableFork` -> `Invalid` when block hash verification fails -
failed hash verification is not a fork-related block issue

etan-status committed:
avoid VC keymanager port conflict in CI (#4459)

`BASE_VC_KEYMANAGER_PORT` was not configured separately between runs,
make it configurable and base on `EXECUTOR_NUMBER` like the others.

tersec committed:
getPayloadV2 support for capella (#4457)

* getPayloadV2 support for capella

* check execution payload type more rigorously

henridf committed:
db.getBlockSZ: Remove no-op 'success' var (#4461)

henridf committed:
EIP4844 gossip transition (#4463)

tersec committed:
add EIP4844 epoch detection for gossip transition tests (#4466)

tersec committed:
spec ref URL & copyright year updates (#4467)

Menduist committed:
Bump libp2p (#4470)

tersec committed:
capella forkchoiceUpdated support (#4462)

* capella forkchoiceUpdated support

* match V2 fcU with V2 getPayload

tersec committed:
bump nim-web3 for correct getPayloadV2 response signature (#4471)

etan-status committed:
re-enable `--sync-light-client` in CI testnets (#4469)

libp2p issues related to operation cancellations have been addressed in
https://github.com/status-im/nim-libp2p/pull/816
This means we can once more enable `--sync-light-client` in CI, without
having to deal with spurious CI failures due to the cancellation issues.

arnetheduck committed:
History pruning (fixes #4419) (#4445)

Introduce (optional) pruning of historical data - a pruned node will
continue to answer queries for historical data up to
`MIN_EPOCHS_FOR_BLOCK_REQUESTS` epochs, or roughly 5 months, capping
typical database usage at around 60-70gb.

To enable pruning, add `--history=prune` to the command line - on the
first start, old data will be cleared (which may take a while) - after
that, data is pruned continuously.

When pruning an existing database, the database will not shrink -
instead, the freed space is recycled as the node continues to run - to
free up space, perform a trusted node sync with a fresh database.

When switching on archive mode in a pruned node, history is retained
from that point onwards.

History pruning is scheduled to be enabled by default in a future
release.

In this PR, `minimal` mode from #4419 is not implemented meaning
retention periods for states and blocks are always the same - depending
on user demand, a future PR may implement `minimal` as well.

tersec committed:
use v1.3.0-rc.0 consensus spec test vectors (#4472)

tersec committed:
revert to Geth stable commits for testnet (#4476)

henridf committed:
Blob storage (#4454)

* Blob storage

* fix indentation

* Fix build (none->Opt.none)

* putBlobs -> putBlobsSidecar

* getBlobs -> getBlobsSidecar

* Check blob correctness when storing a backfill block

* Blobs table: rename and conditionally create

* Check block<->blob match in storeBackfillBlock

* Use when .. toFork() to condition on type

* Check blob viability in block_processor.storeBlock()

* Fix build

* Review feedback

tersec committed:
consensus spec ref URL update v1.3.0-{alpha.2,rc.0}; copyright year update (#4477)

etan-status committed:
disable light client gossip from Capella onward (#4479)

To prevent exchanging incompatible data and therefore disconnect,
disable light client data gossip from Capella onward until specs are ok:
https://github.com/ethereum/consensus-specs/pull/3151

etan-status committed:
add `num_active_participants` helpers for `SyncAggregate` (#4478)

Introduce `num_active_participants` helper function to reduce visibility
of low-level `countOnes` function and reduce code duplication.

tersec committed:
don't use non-engine-API-whitelisted web3 methods (#4484)

etan-status committed:
stricter check for context bytes when recv LC data (#4480)

Bellatrix and Altair light client data share same body, but have other
fork digests. Validate that the peer's sent fork digest matches the one
that we expect (derived from `attested_header.slot`).

tersec committed:
a few consensus spec ref URL updates (#4483)

zah committed:
Bugfix: Handle networks with TTD=0 (e.g. withdrawal devnets)

cheny0 committed:
update based on comments from Zahary and Jacek

cheny0 committed:
Merge branch '4331-run-an-ethereum-node-with-nimbus' of https://github.com/status-im/nimbus-eth2 into 4331-run-an-ethereum-node-with-nimbus

Fix merge conflicts.
